### PR TITLE
 Include more street features such as traffic lights and more to OSM preprocessor output

### LINKED ISSUE #608: STORY: As a handler developer, I want to have access to as much data from OSM as possible, so that I can create more useful audio/haptic map experiences.
- [ ] OSM data: Create list of additional data available from OSM, validate with handler developers
- [ ] OSM data: Extend OSM preprocessor output schema to include new data
- [ ] OSM data: Implement new data in OSM preprocessor
- [ ] OSM data: Test and deploy on unicorn

### PR DESCRIPTION
**Pull request by samueljohnsegun148**
This is in respect to #608.
Please ensure you've followed the checklist and provide all the required information *before* requesting a review.
If you do not have everything applicable to your PR, it will not be reviewed!
If you don't know what something is or if it applies to you, ask!

Don't delete below this line.

---

## Required Information

- [x] I referenced the issue addressed in this PR.
- [x] I described the changes made and how these address the issue.
- [x] I described how I tested these changes.

## Coding/Commit Requirements

* [x] I followed applicable coding standards (e.g., [PEP8](https://pep8.org/))
* [x] I have not committed any models or other large files.

## New Component Checklist (**mandatory** for new microservices)

* [ ] I added an entry to `docker-compose.yml` and `build.yml`.
* [ ] I created A CI workflow under `.github/workflows`.

### CODE REVIEW

**Review by emmanuelwilson — COMMENTED** [review]
If you could check off all of the tasks you have accomplished in the PR, That would be helpful to keep track of which items you have accomplished. Also regarding your test, could you provide context of what are the changes you've made? I like your output but I am not sure what to look for nor can I really know if your changes worked or not! The changes themselves seem good at a glance, I really like your commented code! Keep it up!

**Comment by JRegimbal** [comment]
Can you summarize what the changes here actually do? I can't follow the code and this makes review difficult.

File: preprocessors/openstreetmap/osm_service.py [hunk 1]
```diff
@@ -67,6 +67,7 @@ def server_config2(url, bbox_coord):
     (node({lat_min},{lon_min},{lat_max},{lon_max}) ["amenity"];
     way({lat_min},{lon_min},{lat_max},{lon_max}) ["amenity"];
     rel({lat_min},{lon_min},{lat_max},{lon_max}) ["amenity"];
+    way({lat_min},{lon_min},{lat_max},{lon_max}) ["building"];
```
Comment (samueljohnsegun148) [inline preprocessors/openstreetmap/osm_service.py:70]: This line queries other features (POIs) that may not be available under the ```amenity``` key.

File: preprocessors/openstreetmap/osm_service.py [hunk 2]
```diff
@@ -122,17 +123,14 @@ def process_streets_data(OSM_data, bbox_coordinates):
                     "lat": float(node.lat),
                     "lon": float(node.lon),
                 }
+                # Include tags for a street node if available
+                node_object.update(node.tags)
```
Comment (samueljohnsegun148) [inline preprocessors/openstreetmap/osm_service.py:127]: While POIs under the ```amenity``` and ```building``` keys are off-route landmarks, this specifically ensures that attributes are extracted for nodes of a way if available and this is where we can get information on the traffic, crossing, etc.

File: preprocessors/openstreetmap/osm_service.py [hunk 3]
```diff
@@ -604,7 +611,9 @@ def allot_intersection(processed_OSM_data, inters_rec_up
                     if id1 != id2:  # compare unique street only
                         # check if a node represents an intersection
                         if nodes[i] == intersection_nodes[items]:
-                            nodes[i]["cat"] = "intersection"
```
Comment (samueljohnsegun148) [inline preprocessors/openstreetmap/osm_service.py:607]: This was changed based on the discussion we had to include ```ids``` of intersecting streets instead of indicating if there is an intersection or not.

**Comment by samueljohnsegun148** [comment]
Basically, the changes made are on the points of interest. For an example, a new key ```intersecttion``` is now being used to hold the ids of the intersecting streets. This is different from the existing case of ```cat: intersection``` type, following our latest discussion. Information such as traffic, crossing, tactile paving, etc.,  is also now included. Generally, landmarks (POIs) and also the streets, with these changes, currently have more attributes, as much as is available in the OpenStreetMap. 

```
{
				"id": 208765235,
				"lat": 45.5060216,
				"lon": -73.5709576,
				"highway": "traffic_signals",
				"cat": "traffic_signals",
				"intersection": [
					19896822,
					465414042
				],
				"name": "Rue City-Councillors intersecting Avenue du Président-Kennedy"
			},
{
				"id": 6912952956,
				"lat": 45.5052563,
				"lon": -73.5706173,
				"crossing": "marked",
				"highway": "crossing",
				"tactile_paving": "yes",
				"cat": "crossing",
				"intersection": [
					738275183,
					465416568
				],
				"name": "footway intersecting Boulevard De Maisonneuve Ouest"
			}
```
You may want to check though this [file attached](https://github.com/Shared-Reality-Lab/IMAGE-server/files/11520730/outputfile.txt)

**Comment by JRegimbal** [comment]
This seems reasonable, but really this should have been several different pull requests. Only the inclusion of traffic lights, lanes, etc. are relevant to the linked issue.

File: preprocessors/openstreetmap/osm_service.py (hunk 1 above)
Comment (JRegimbal) [reply preprocessors/openstreetmap/osm_service.py:70]: Could you give an example of one such feature?

File: preprocessors/openstreetmap/osm_service.py [hunk 4]
```diff
@@ -159,18 +157,27 @@ def process_streets_data(OSM_data, bbox_coordinates):
                     "street_id": int(way.id),
                     "street_name": way.tags.get("name"),
                     "street_type": way.tags.get("highway"),
-                    "addr:street": way.tags.get("addr:street"),
-                    "surface": way.tags.get("surface"),
                     "oneway": oneway,
-                    "sidewalk": way.tags.get("sidewalk"),
-                    "maxspeed": way.tags.get("maxspeed"),
```
Comment (JRegimbal) [inline preprocessors/openstreetmap/osm_service.py:166]: Why remove sidewalk, maxspeed, and surface?

File: preprocessors/openstreetmap/osm_service.py (hunk 3 above)
Comment (JRegimbal) [reply preprocessors/openstreetmap/osm_service.py:607]: Understandable, but why is it in *this* pull request?

**Review by JRegimbal — COMMENTED** [review]
Few points of clarification

File: preprocessors/openstreetmap/osm_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py:70]: The ```building``` key contains almost the same information on the points of interest as the ```amenity``` key, but none of the two has complete data, hence the inclusion in case any information is available in one but missing in the other.

File: preprocessors/openstreetmap/osm_service.py (hunk 4 above)
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py:166]: They are not removed, I simply found a better way to extract them together with the new features in request. See a sample below, which is also in the file attached above:
```"street_id": 884573684,
				"street_name": "Boulevard De Maisonneuve Ouest",
				"street_type": "tertiary",
				"oneway": true,
				"cycleway:left": "track",
				"cycleway:left:oneway": "no",
				"oneway:bicycle": "no",
				"surface": "asphalt",
				"nodes": [
					{
						"id": 7181949559,
						"lat": 45.506639321605924,
						"lon": -73.56938825256503,
						"node_type": "displaced"
					},
					{
						"id": 1201312550,
						"lat": 45.5065869,
						"lon": -73.5694346,
						"POIs_ID": [
							2479541440
						]
					},
					{
						"id": 8577070630,
						"lat": 45.5058039,
						"lon": -73.5701836
					},
					{
						"id": 208756317,
						"lat": 45.505699,
						"lon": -73.5702839,
						"POIs_ID": [
							208756317
						]
					}
				]
			},
```

**Comment by samueljohnsegun148** [comment]
The challenge of representing some of the features (e.g. where we have traffic lights, crossings, intersections, etc, as tags for a particular node) came up after extracting some of the information, which specifically prompted the meeting we had. So, it is very important that this change comes along because what is extracted also has to be represented. I equally ran this with Vennissa. A few more PRs are readily prepared. Just only waiting for this to get approved.

File: preprocessors/openstreetmap/osm_service.py (hunk 1 above)
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py:70]: For an example, hotels are supposed to be under the ```amenity``` key, but this was only found under the ```building``` key. There are many others like that, depending on the queried area.
```{
				"id": 341125395,
				"lat": 45.5139393,
				"lon": -73.557474,
				"name": "Hôtel Le Roberval",
				"cat": "building",
				"tourism": "hotel"
			},```
[Full file attached](https://github.com/Shared-Reality-Lab/IMAGE-server/files/11525773/fullfile.txt) using with the request below:
```
{
  "request_uuid": "a3d7b6be-8f3b-47ba-8b17-4b0557e7a8ce",
  "timestamp": 1637789517,
	"coordinates": {
     "latitude": 45.5140079,	
     "longitude":  -73.5576234
   },
  "context": "",
  "language": "en",
  "url": "https://fake.site.com/some-url",
  "capabilities": ["ca.mcgill.a11y.image.capability.DebugMode"],
  "renderers": [
    "ca.mcgill.a11y.image.renderer.Text",
    "ca.mcgill.a11y.image.renderer.SimpleAudio",
    "ca.mcgill.a11y.image.renderer.SegmentAudio",
		"ca.mcgill.a11y.image.renderer.SVGLayers"
  ]
}
```

Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py:70]: For an example, hotels are supposed to be under the ```amenity``` key, but this was only found under the ```building``` key. There are many others like that, depending on the queried area.
```{
				"id": 341125395,
				"lat": 45.5139393,
				"lon": -73.557474,
				"name": "Hôtel Le Roberval",
				"cat": "building",
				"tourism": "hotel"
			},
```
[Full file attached](https://github.com/Shared-Reality-Lab/IMAGE-server/files/11525773/fullfile.txt) using with the request below:
```
{
  "request_uuid": "a3d7b6be-8f3b-47ba-8b17-4b0557e7a8ce",
  "timestamp": 1637789517,
	"coordinates": {
     "latitude": 45.5140079,	
     "longitude":  -73.5576234
   },
  "context": "",
  "language": "en",
  "url": "https://fake.site.com/some-url",
  "capabilities": ["ca.mcgill.a11y.image.capability.DebugMode"],
  "renderers": [
    "ca.mcgill.a11y.image.renderer.Text",
    "ca.mcgill.a11y.image.renderer.SimpleAudio",
    "ca.mcgill.a11y.image.renderer.SegmentAudio",
		"ca.mcgill.a11y.image.renderer.SVGLayers"
  ]
}
```

**Review by JRegimbal — APPROVED** [review]
Several parts _should_ be present in other pull requests because they don't seem to address the issue linked here, but good for now. Be careful in the future though as I *will* be refusing PRs that are not properly scoped to one problem/feature.
